### PR TITLE
Deploy contracts through generic factory

### PR DESCRIPTION
**Pull request by jeremyjams**
Done in this PR:
1.  Deploy contracts through generic factory:
```
$ npx hardhat deploy
[...]
ChainId: 31337n
Deploying all contracts related to voucher..
Using admin address: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
Using manager address: 0x70997970C51812dc3A010C7d01b50e0d17dc79C8
Using minter address: 0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC
Using PoCo address: 0x123456789a123456789b123456789b123456789d
Deploying factory on this network..
Factory: 0xfAC000a12dA42B871c0AaD5F25391aAe62958Db1
Salt: 0x0000000000000000000000000000000000000000000000000000000000000000
VoucherImpl: 0x3a009D350b8C6851BD9A0Dcb1352CBC5840a6096 
VoucherUpgradeableBeacon: 0x97A48709497776f58C1B4280D06BbCa5f3Cdf7d7 
VoucherHubImpl: 0x8940F91c2020286E1A1698616ca86DB2c4CB8F3C 
VoucherHubERC1967Proxy: 0x3eE5875781Da14c1917871dF912C2b69f207d33d
```
2. Validate deployment with tests
```
$ npm run test
[...]
  VoucherHub
    Initialize
Deploying all contracts related to voucher..
Using admin address: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
Using manager address: 0x70997970C51812dc3A010C7d01b50e0d17dc79C8
Using minter address: 0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC
Using PoCo address: 0x5FbDB2315678afecb367f032d93F642f64180aa3
Deploying factory on this network..
Factory: 0xfAC000a12dA42B871c0AaD5F25391aAe62958Db1
Salt: 0x0000000000000000000000000000000000000000000000000000000000000000
[...]

  106 passing (6s)
```

### CODE REVIEW

File: contracts/VoucherHub.sol [hunk 1]
```diff
@@ -66,14 +66,15 @@ contract VoucherHub is
     }
 
     function initialize(
```
Comment (jeremyjams) [inline contracts/VoucherHub.sol:68]: Eventually notify clients before merging

File: test/VoucherHub.test.ts [hunk 2]
```diff
@@ -84,9 +86,12 @@ describe('VoucherHub', function () {
                 .to.equal(await voucherHub.defaultAdmin())
                 .to.equal(admin);
             expect(await voucherHub.defaultAdminDelay()).to.equal(0);
-            expect(await voucherHub.hasRole(await voucherHub.UPGRADER_ROLE.staticCall(), admin));
-            expect(await voucherHub.hasRole(await voucherHub.MANAGER_ROLE.staticCall(), manager));
-            expect(await voucherHub.hasRole(await voucherHub.MINTER_ROLE.staticCall(), minter));
+            expect(await voucherHub.hasRole(await voucherHub.UPGRADER_ROLE.staticCall(), admin)).to
+                .be.true;
```
Comment (jeremyjams) [inline test/VoucherHub.test.ts:90]: expect was no working before this `to.be.true` update

File: package.json [hunk 3]
```diff
@@ -17,6 +17,7 @@
     "prepare": "husky",
     "build": "npx hardhat compile",
     "test": "npx hardhat test",
+    "test-ci": "npm run test && FACTORY=false npm run test",
```
Comment (jeremyjams) [inline package.json:20]: Target is to run tests with contracts deployed through factory **and** through EOA (at least for now)

File: contracts/VoucherHub.sol (hunk 1 above)
Comment (gfournieriExec) [reply contracts/VoucherHub.sol:68]: Yes good point, hopefully that won't break their work, it's just the ABI that change but they never call this function

File: test/VoucherHub.test.ts (hunk 2 above)
Comment (gfournieriExec) [reply test/VoucherHub.test.ts:90]: why did it work before then ?

**Review by gfournieriExec — APPROVED** [review]

File: test/VoucherHub.test.ts (hunk 2 above)
Comment (jeremyjams) [reply test/VoucherHub.test.ts:90]: It was not really checking, it was more like console.log() ..!

File: deploy/deploy.ts [hunk 4]
```diff
@@ -1,36 +1,138 @@
...
+    GenericFactory_shanghai__factory,
+    UpgradeableBeacon,
+    UpgradeableBeacon__factory,
+    VoucherHub__factory,
+    Voucher__factory,
+} from '../typechain-types';
 
 export default async function (hre: HardhatRuntimeEnvironment) {
     const chainId = (await ethers.provider.getNetwork()).chainId;
     console.log('ChainId:', chainId);
```
[the remark this answers is not in the document]
Comment (jeremyjams) [reply deploy/deploy.ts]: Possible yes, see [Deploy with explicit poco address or get from config](https://github.com/iExecBlockchainComputing/iexec-voucher-contracts/pull/35/commits/a14cf7cdeceed25dfb06825281ae361ad61943da)

File: deploy/deploy.ts [hunk 5]
```diff
@@ -46,24 +148,26 @@ async function deployAll(
...
-    let pocoAddress: string = deploymentConfig[chainId]?.pocoAddress;
+    const config = deploymentConfig[chainId];
+    let pocoAddress: string = config?.pocoAddress;
     // Override config if required.
     if (process.env.IEXEC_POCO_ADDRESS) {
-        pocoAddress = process.env.IEXEC_POCO_ADDRESS;
+        config.pocoAddress = process.env.IEXEC_POCO_ADDRESS;
     }
     // Check final config.
     if (!ethers.isAddress(pocoAddress)) {
```
Comment (zguesmi) [inline deploy/deploy.ts:174]:
```suggestion
    if (!ethers.isAddress(config.pocoAddress)) {
```

File: deploy/deploy.ts [hunk 6]
```diff
@@ -46,24 +148,26 @@ async function deployAll(
...
 }
 
 /**
  * Get deployment config according to chain.
  */
 async function getDeploymentConfig(chainId: number) {
     // Read default config of the target chain.
-    let pocoAddress: string = deploymentConfig[chainId]?.pocoAddress;
+    const config = deploymentConfig[chainId];
+    let pocoAddress: string = config?.pocoAddress;
```
Comment (zguesmi) [inline deploy/deploy.ts:168]:
```suggestion
```

File: deploy/deploy.ts (hunk 5 above)
Comment (jeremyjams) [reply deploy/deploy.ts]: Thanks! For all see [Verify poco address is set](https://github.com/iExecBlockchainComputing/iexec-voucher-contracts/pull/35/commits/9e90c907c7cf72af3fc8806399521f6a2180a1cf)

**Review by zguesmi — APPROVED** [review]
Neat!

**Comment by zguesmi** [comment]
FYI @PierreJeanjacquot because deployment files has been updated.

**Pull request merged**